src/msg_recorder/recorder.h

```diff
@@ -25,7 +25,7 @@ class MessageRecorder : public CRNamedNode<MessageRecorder> {
     MessageRecorder& operator=(const MessageRecorder&) = delete;
     MessageRecorder& operator=(MessageRecorder&&) = delete;
 
-    ~MessageRecorder();
+    ~MessageRecorder() override;
```

Comment:
我建议带override的头部都加上virtual

Reply:
virtual override没意义，virtual是多余的。不virtual的写override本身就是warning。而且我记得有readability check把virtual override算作过于啰嗦的，我不确定我们有没有开但我觉得没必要再加virtual

Reply:
很有意义，因为override才是废话的那部分，是作为一个virtual之上的检查，叫compiler验证一下基类有virtual。如果没有virtual，override一旦因为任何原因被删就会改变含义。

Reply:
- 我们有这条检查 https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/explicit-virtual-functions.html
    - ` error: 'virtual' is redundant since the function is already declared 'override' [cppcoreguidelines-explicit-virtual-functions,-warnings-as-errors]`

> 很有意义，因为override才是废话的那部分，是作为一个virtual之上的检查，叫compiler验证一下基类有virtual。

对，但这并不是废话。override一个不virtual的函数是error，override的函数签名不对也是error。所以override是比virtual更强的写法，而virtual是多余的。

> 如果没有virtual，override一旦因为任何原因被删就会改变含义。比如B如果去掉了基类A，就需要去掉override，但有可能这时候会忘记加virtual，导致它根本不在vtable里。如果B有个派生类C，C里写的virtual的含义就会被这个影响。而override的作用是，如果从C的角度看，这时候可以靠override去检查B有没有virtual。

那这样C的override就会报错了。因为不在vtable里。所以我觉得正确的做法是该override必须override（这个检查很基础，默认就会开clang-diagnostic-inconsistent-missing-override），这样既更对又不啰嗦。

Reply:
那就去掉这个check直到它能允许override和virtual共存，或者至少不要werror，它这里的设计我觉得有问题。doc里说对于compiler来说，有override后virtual就是redundant的，但是对人不是。

> 对，但这并不是废话。override一个不virtual的函数是error，override的函数签名不对也是error。所以override是比virtual更强的写法，而virtual是多余的。

编译方面virtual是override的超集，override不能用于表明第一个virtual是virtual，而派生类里可以不写virtual并不是因为override有generate vtable的含义，这个含义是从virtual本身会被自动继承来的。只是因为override要求了基类有virtual才碰巧导致了它能imply virtual

Reply:
> 但是对人不是。

为什么不是，对人来说一样不需要过多同样含义的标记，这样 1. 难读 2. 容易导致无意义的修改

> 编译方面virtual是override的超集

并不是，他们只是有交集

> override不能用于表明第一个virtual（基类的）是virtual

对，但你本来也不能在基类写override

> 而派生类里可以不写virtual并不是因为override有generate vtable的含义，这个含义是从virtual本身会被自动继承来的

是，所以语义上可以使用override的地方，本来就**必须**是virtual的。但virtual不意味着override。所以在这种情况下，override包含virtual。

只在基类用virtual的好处是
- 减少冗余标记
- 清楚区分基类和override的函数

我没有觉得有什么缺点。因为编译器警告+tidy可以帮我们检查：
- override必须在基类有virtual
- 有override必须写出来

Reply:
> 为什么不是，对人来说一样不需要过多同样含义的标记，这样 1. 难读 2. 容易导致无意义的修改

因为上面说的，如果基类去掉virtual的话，去去掉派生类里override的人很可能忘记补上virtual。
而且有一些特殊的场合，比如存在多个版本的upstream API，有的虚有的不虚，但继承时想要无条件让它虚，就需要用只写virtual不写override的组合。这个是作者才知道但编译器不知道的信息。

> 并不是，他们只是有交集
> 对，但你本来也不能在基类写override
> 是，所以语义上可以使用override的地方，本来就必须是virtual的。但virtual不意味着override。所以在这种情况下，override包含virtual。

我这里指的编译是codegen，不是semantic check。
override只报错，不改变codegen输出，它能扯上vtable完全是因为第一层基类的virtual

> 清楚区分基类和override的函数

上面说了，存在应该只用virtual的派生类

Reply:
> 因为上面说的，如果基类去掉virtual的话，去去掉派生类里override的人很可能忘记补上virtual。

这样有两个问题

- review时很容易看出来
- 如果已经有子类，不补上virtual是error

> 而且有一些特殊的场合，比如存在多个版本的upstream API，有的虚有的不虚，但继承时想要无条件让它虚，就需要用只写virtual不写override的组合。这个是作者才知道但编译器不知道的信息。

这种情况我目前没遇到过，可是真要是这样你总是可以nolint+解释的，没必要为这种少见的情况影响其他地方。

> override只报错，不改变codegen输出，它能扯上vtable完全是因为第一层基类的virtual

对人来说，这个语义哪里来的没什么区别。

> > 清楚区分基类和override的函数
> 
> 上面说了，存在应该只用virtual的派生类

我觉得这种情况不会很多，出现了就nolint掉就完了，正好还强制你解释这种复杂情况

Reply:
> review时很容易看出来

我遇到过很多次这种坑。
有我自己干的，也有review漏过后来抓到的。

> 如果已经有子类，不补上virtual是error

是正常的overwrite，什么error?

> 这种情况我目前没遇到过，可是真要是这样你总是可以nolint+解释的，没必要为这种少见的情况影响其他地方。

这个我是认可的。

> 对人来说，这个语义哪里来的没什么区别。

Generally speaking implied semantics can often has hidden side effects or corner cases.
Every time you want to prove something you'll have to think again whether all cases are covered.

And the implication itself is not future-proof.
Newly added rules can easily alter the result.
That's just a simple difference between axiom and theorem.

Reply:
Per offline discussion, we will turn the [explicit virtual function check](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/explicit-virtual-functions.html) on and disallow redundant `virtual` for now.